File: backend/routers/rpki/rpki.py
"""RPKI Protocol Router.

API endpoints for managing VyOS RPKI (Resource Public Key Infrastructure)
configuration. Supports version-aware configuration for VyOS 1.4 and 1.5.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import RpkiBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/rpki", tags=["rpki"])


# ============================================================================
# Pydantic Models
# ============================================================================


class RpkiSshConfig(BaseModel):
    key: Optional[str] = None
    username: Optional[str] = None


class RpkiCacheServer(BaseModel):
    address: str
    port: Optional[int] = None
    preference: Optional[int] = None
    source_address: Optional[str] = None
    ssh: Optional[RpkiSshConfig] = None


class RpkiConfig(BaseModel):
    cache_servers: List[RpkiCacheServer] = []
    expire_interval: Optional[int] = None
    polling_period: Optional[int] = None
    retry_interval: Optional[int] = None


class RpkiBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(None, description="Comma-separated arguments")


class RpkiBatchRequest(BaseModel):
    operations: List[RpkiBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_rpki_capabilities(request: Request):
    """Return RPKI feature capabilities based on the device VyOS version."""
    await require_read_permission(request, FeatureGroup.RPKI)

    try:
        service = get_session_vyos_service(request)
        builder = RpkiBatchBuilder(version=service.get_version())
        capabilities = builder.get_capabilities()

        if hasattr(request.state, "instance") and request.state.instance:
            capabilities["instance_name"] = request.state.instance.get("name")
            capabilities["instance_id"] = request.state.instance.get("id")

        return capabilities
    except Exception:
        logger.exception("Unhandled error in get_rpki_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=RpkiConfig)
async def get_rpki_config(http_request: Request, refresh: bool = False):
    """Return the full RPKI configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.RPKI)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        rpki_raw = full_config.get("protocols", {}).get("rpki", {})

        if not rpki_raw:
            return RpkiConfig()

        return RpkiConfig(
            cache_servers=_parse_cache_servers(rpki_raw.get("cache", {})),
            expire_interval=_safe_int(rpki_raw.get("expire-interval")),
            polling_period=_safe_int(rpki_raw.get("polling-period")),
            retry_interval=_safe_int(rpki_raw.get("retry-interval")),
        )
    except Exception:
        logger.exception("Unhandled error in get_rpki_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config Parsers
# ============================================================================


def _safe_int(value) -> Optional[int]:
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def _parse_cache_servers(raw: dict) -> List[RpkiCacheServer]:
    if not raw:
        return []

    servers = []
    for address, cfg in raw.items():
        if cfg is None:
            cfg = {}

        ssh_raw = cfg.get("ssh", {}) or {}
        ssh = None
        if ssh_raw:
            ssh = RpkiSshConfig(
                key=ssh_raw.get("key"),
                username=ssh_raw.get("username"),
            )

        servers.append(RpkiCacheServer(
            address=address,
            port=_safe_int(cfg.get("port")),
            preference=_safe_int(cfg.get("preference")),
            source_address=cfg.get("source-address"),
            ssh=ssh,
        ))

    return servers


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def rpki_batch_configure(http_request: Request, body: RpkiBatchRequest):
    """Execute a batch of RPKI configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.RPKI)

    try:
        service = get_session_vyos_service(http_request)
        builder = RpkiBatchBuilder(version=service.get_version())

        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
            elif len(params) == 2 and operation.value is not None:
                values = operation.value.split(",", 1)
                if len(values) == 2:
                    method(values[0], values[1])
                else:
                    method(operation.value, "")

        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "RPKI configuration updated"},
            error=response.error if response.error else None,
        )
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {str(e)}")
    except Exception:
        logger.exception("Unhandled error in rpki_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")

File: frontend/src/lib/api/rpki.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface RpkiSshConfig {
  key: string | null;
  username: string | null;
}

export interface RpkiCacheServer {
  address: string;
  port: number | null;
  preference: number | null;
  source_address: string | null;
  ssh: RpkiSshConfig | null;
}

export interface RpkiConfig {
  cache_servers: RpkiCacheServer[];
  expire_interval: number | null;
  polling_period: number | null;
  retry_interval: number | null;
}

export interface RpkiCapabilities {
  version: string;
  version_info: { is_1_4: boolean; is_1_5: boolean };
  features: {
    cache_servers: { supported: boolean; description: string };
    expire_interval: { supported: boolean; description: string };
    polling_period: { supported: boolean; description: string };
    retry_interval: { supported: boolean; description: string };
    ssh_transport: { supported: boolean; description: string };
  };
  instance_name?: string;
  instance_id?: string;
}

export interface RpkiBatchOperation {
  op: string;
  value?: string;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// API Service
// ============================================================================

class RpkiService {
  async getCapabilities(): Promise<RpkiCapabilities> {
    return apiClient.get<RpkiCapabilities>("/vyos/rpki/capabilities");
  }

  async getConfig(refresh = false): Promise<RpkiConfig> {
    return apiClient.get<RpkiConfig>("/vyos/rpki/config", {
      refresh: refresh.toString(),
    });
  }

  private async batch(ops: RpkiBatchOperation[]): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/rpki/batch", {
      operations: ops,
    });
    if (!result.success) throw new Error(result.error || "RPKI operation failed");
    return result;
  }

  async createCacheServer(server: RpkiCacheServer): Promise<VyOSResponse> {
    const ops: RpkiBatchOperation[] = [{ op: "set_cache", value: server.address }];

    if (server.port != null) {
      ops.push({ op: "set_cache_port", value: `${server.address},${server.port}` });
    }
    if (server.preference != null) {
      ops.push({ op: "set_cache_preference", value: `${server.address},${server.preference}` });
    }
    if (server.source_address) {
      ops.push({ op: "set_cache_source_address", value: `${server.address},${server.source_address}` });
    }
    if (server.ssh?.key) {
      ops.push({ op: "set_cache_ssh_key", value: `${server.address},${server.ssh.key}` });
    }
    if (server.ssh?.username) {
      ops.push({ op: "set_cache_ssh_username", value: `${server.address},${server.ssh.username}` });
    }

    return this.batch(ops);
  }

  async updateCacheServer(
    original: RpkiCacheServer,
    updated: RpkiCacheServer
  ): Promise<VyOSResponse> {
    const ops: RpkiBatchOperation[] = [];
    const addr = original.address;

    if (original.port !== updated.port) {
      if (updated.port != null) {
        ops.push({ op: "set_cache_port", value: `${addr},${updated.port}` });
      } else {
        ops.push({ op: "delete_cache_port", value: addr });
      }
    }

    if (original.preference !== updated.preference) {
      if (updated.preference != null) {
        ops.push({ op: "set_cache_preference", value: `${addr},${updated.preference}` });
      } else {
        ops.push({ op: "delete_cache_preference", value: addr });
      }
    }

    if (original.source_address !== updated.source_address) {
      if (updated.source_address) {
        ops.push({ op: "set_cache_source_address", value: `${addr},${updated.source_address}` });
      } else {
        ops.push({ op: "delete_cache_source_address", value: addr });
      }
    }

    const hadSsh = original.ssh != null;
    const hasSsh = updated.ssh != null;

    if (hadSsh && !hasSsh) {
      ops.push({ op: "delete_cache_ssh", value: addr });
    } else {
      if (original.ssh?.key !== updated.ssh?.key) {
        if (updated.ssh?.key) {
          ops.push({ op: "set_cache_ssh_key", value: `${addr},${updated.ssh.key}` });
        } else {
          ops.push({ op: "delete_cache_ssh_key", value: addr });
        }
      }
      if (original.ssh?.username !== updated.ssh?.username) {
        if (updated.ssh?.username) {
          ops.push({ op: "set_cache_ssh_username", value: `${addr},${updated.ssh.username}` });
        } else {
          ops.push({ op: "delete_cache_ssh_username", value: addr });
        }
      }
    }

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batch(ops);
  }

  async deleteCacheServer(address: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_cache", value: address }]);
  }

  async updateGlobalSettings(
    current: Pick<RpkiConfig, "expire_interval" | "polling_period" | "retry_interval">,
    next: Pick<RpkiConfig, "expire_interval" | "polling_period" | "retry_interval">
  ): Promise<VyOSResponse> {
    const ops: RpkiBatchOperation[] = [];

    if (current.expire_interval !== next.expire_interval) {
      if (next.expire_interval != null) {
        ops.push({ op: "set_expire_interval", value: String(next.expire_interval) });
      } else {
        ops.push({ op: "delete_expire_interval" });
      }
    }

    if (current.polling_period !== next.polling_period) {
      if (next.polling_period != null) {
        ops.push({ op: "set_polling_period", value: String(next.polling_period) });
      } else {
        ops.push({ op: "delete_polling_period" });
      }
    }

    if (current.retry_interval !== next.retry_interval) {
      if (next.retry_interval != null) {
        ops.push({ op: "set_retry_interval", value: String(next.retry_interval) });
      } else {
        ops.push({ op: "delete_retry_interval" });
      }
    }

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batch(ops);
  }
}

export const rpkiService = new RpkiService();
